internal/websocket/ws.go

```diff
@@ -284,7 +289,21 @@ func (ws *Websocket) run(ctx context.Context) {
 			// Read loop
 			for {
 				var msg wrp.Message
+				ctx, cancel := context.WithTimeout(ctx, ws.inactivityTimeout)
 				typ, reader, err := ws.conn.Reader(ctx)
+				if errors.Is(err, context.DeadlineExceeded) {
+					select {
+					case <-inactivityTimeout:
```

Comment:
You don't need a channel created via `time.After` here.  Since you've just got a single loop, you can use a (possibly atomic) timestamp that is the expiry time.  Something like:

```
expiry := time.Now().Add( /* whatever the timeout is */ )
for {
    typ, reader, err := ws.conn.Reader(ctx)
    if errors.Is(err, context.DeadlineExceeded) {
        if time.Now().After(expiry) {
             // timeout exceeded
        }

        // new expiry for the next iteration
        expiry = expiry.Add( /* timeout */ )
    }
}
```

If other sources need to update the expiry, then pull it out as a member on the `Websocket` type and treat it as an atomic variable, using the unix timestamp value:

```
type Websocket struct {
    expiry atomic.Int64 // use the new API, it's simpler
}

func (ws *Websocket) Start() {
    // you don't have to use UTC, but it makes debugging easier since no matter where this
    // code is run, the expiry will be in the same TZ
    ws.expiry.Store(time.Now().UTC().Unix())

    // continue the for loop as above, but use ws.expiry.Load and ws.expiry.Store to access the expiry value
}
```

The above allows `Websocket` to have a method that updates the expiry value in a lock-free way that is concurrent-safe.

Reply:
great point! I choose `time.After` since we're already using one in the ending context (trying to be consistent). But that's a different use case where `time.After` makes more sense. I can see how `time.Now().Add()` would fit better here and provide us the opportunity to expose it safely if we ever wanted to :)


```go
select {  
case <-time.After(next):  
case <-ctx.Done():  
    return  
}
```

Reply:
@schmidtw do you have a preference?

Reply:
Also, to make this kind of code testable, you don't want to use `time.Now` directly.  Instead, use a `func() time.Time` and you inject into the `Websocket` type.  This is why I created github.com/xmidt-org/chronon.  It makes testing time-related logic easier.

Reply:
I'll implement this in a new pr 🙂